File: client/src/main/java/io/littlehorse/jlib/worker/internal/util/VariableMapping.java
package io.littlehorse.jlib.worker.internal.util;

import com.fasterxml.jackson.core.JsonProcessingException;
import io.littlehorse.jlib.common.LHLibUtil;
import io.littlehorse.jlib.common.exception.InputVarSubstitutionError;
import io.littlehorse.jlib.common.exception.TaskSchemaMismatchError;
import io.littlehorse.jlib.common.proto.ScheduledTaskPb;
import io.littlehorse.jlib.common.proto.TaskDefPb;
import io.littlehorse.jlib.common.proto.VarNameAndValPb;
import io.littlehorse.jlib.common.proto.VariableDefPb;
import io.littlehorse.jlib.common.proto.VariableValuePb;
import io.littlehorse.jlib.worker.WorkerContext;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class VariableMapping {

    private String name;
    private Class<?> type;
    private int position;

    public VariableMapping(
        TaskDefPb taskDef,
        int position,
        Class<?> type,
        String javaParamName
    ) throws TaskSchemaMismatchError {
        this.type = type;

        if (type.equals(WorkerContext.class)) return;
        this.position = position;

        if (position >= taskDef.getInputVarsCount()) {
            throw new TaskSchemaMismatchError(
                "Provided Java function has more parameters than the TaskDef."
            );
        }
        this.name = javaParamName;
        VariableDefPb input = taskDef.getInputVars(position);

        String msg = null;

        switch (input.getType()) {
            case INT:
                if (!LHLibUtil.isINT(type)) {
                    msg = "TaskDef provides INT, func accepts " + type.getName();
                }
                break;
            case DOUBLE:
                if (!LHLibUtil.isDOUBLE(type)) {
                    msg = "TaskDef provides a DOUBLE, func accepts " + type.getName();
                }
                break;
            case STR:
                if (!LHLibUtil.isSTR(type)) {
                    msg = "TaskDef provides a STRING, func accepts " + type.getName();
                }
                break;
            case BOOL:
                if (!LHLibUtil.isBOOL(type)) {
                    msg = "TaskDef provides a BOOL, func accepts " + type.getName();
                }
                break;
            case BYTES:
                if (!LHLibUtil.isBYTES(type)) {
                    msg = "TaskDef provides BYTES, func accepts " + type.getName();
                }
                break;
            case JSON_ARR:
            case JSON_OBJ:
                log.info(
                    "Info: Will use Jackson to deserialize Json into {}",
                    type.getName()
                );
                break;
            case NULL:
            case UNRECOGNIZED:
                throw new RuntimeException("Not possible");
        }

        if (msg != null) {
            throw new TaskSchemaMismatchError(
                "Invalid assignment for var " + name + ": " + msg
            );
        }
    }

    public Object assign(ScheduledTaskPb taskInstance, WorkerContext context)
        throws InputVarSubstitutionError {
        if (type.equals(WorkerContext.class)) {
            return context;
        }

        VarNameAndValPb assignment = taskInstance.getVariables(position);
        String taskDefParamName = assignment.getVarName();
        VariableValuePb val = assignment.getValue();

        String jsonStr = null;

        // We've already done validation for the
        switch (val.getType()) {
            case INT:
                if (type == Long.class || type == long.class) {
                    return val.getInt();
                } else {
                    return (int) val.getInt();
                }
            case DOUBLE:
                if (type == Double.class || type == double.class) {
                    return val.getDouble();
                } else {
                    return (float) val.getDouble();
                }
            case STR:
                return val.getStr();
            case BYTES:
                return val.getBytes().toByteArray();
            case BOOL:
                return val.getBool();
            case JSON_ARR:
                jsonStr = val.getJsonArr();
                break;
            case NULL:
                return null;
            case JSON_OBJ:
                jsonStr = val.getJsonObj();
                break;
            case UNRECOGNIZED:
                throw new RuntimeException("Unrecognized variable value type");
        }

        try {
            return LHLibUtil.deserializeFromjson(jsonStr, type);
        } catch (JsonProcessingException exn) {
            throw new InputVarSubstitutionError(
                "Failed deserializing the Java object for variable " +
                taskDefParamName,
                exn
            );
        }
    }
}

File: sdk-dotnet/LittleHorseSDK/LittleHorse.Worker/Internal/VariableMapping.cs
using LittleHorse.Common.Exceptions;
using LittleHorse.Worker.Internal.Helpers;
using LittleHorseSDK.Common.proto;
using Microsoft.Extensions.Logging;

namespace LittleHorse.Worker.Internal
{
    public class VariableMapping
    {
        private ILogger? _logger;
        private string? _name;
        private Type _type;
        private int _position;

        public VariableMapping(TaskDef taskDef, int position, Type type, string? paramName, ILogger? logger = null)
        {
            _type = type;
            _name = paramName;
            _position = position;
            _logger = logger;

            if (_type.IsAssignableFrom(typeof(LHWorkerContext)))
            {
                return;
            }

            var input = taskDef.InputVars[position];

            ValidateType(input.Type, _type, _name);
        }

        public object? Assign(ScheduledTask taskInstance, LHWorkerContext workerContext)
        {
            if (_type.GetType() == typeof(LHWorkerContext))
            {
                return workerContext;
            }

            VarNameAndVal assignment = taskInstance.Variables[_position];
            string taskDefParamName = assignment.VarName;
            VariableValue val = assignment.Value;

            string? jsonStr;

            switch (val.Type)
            {
                case VariableType.Int:
                    if (_type == typeof(long) || _type == typeof(long?))
                    {
                        return val.Int;
                    }
                    else
                    {
                        return (int)val.Int;
                    }
                case VariableType.Double:
                    if (_type == typeof(double) || _type == typeof(double?))
                    {
                        return val.Double;
                    }
                    else
                    {
                        return (float)val.Double;
                    }
                case VariableType.Str:
                    return val.Str;
                case VariableType.Bytes:
                    return val.Bytes.ToByteArray();
                case VariableType.Bool:
                    return val.Bool;
                case VariableType.JsonArr:
                    jsonStr = val.JsonArr;
                    break;
                case VariableType.Null:
                    return null;
                case VariableType.JsonObj:
                    jsonStr = val.JsonObj;
                    break;
                default:
                    throw new InvalidOperationException("Unrecognized variable value type");
            }

            try
            {
                return LHMappingHelper.DeserializeFromJson(jsonStr, _type);
            }
            catch (Exception ex)
            {
                throw new LHInputVarSubstitutionException($"Failed deserializing the C# object for variable {taskDefParamName}", ex);
            }
        }


        private void ValidateType(VariableType taskDefInputType, Type paramType, string? paramName)
        {
            string errorMsg = string.Empty;

            switch (taskDefInputType)
            {
                case VariableType.Int:
                    if (!paramType.IsAssignableFrom(typeof(int)))
                    {
                        errorMsg = $"TaskDef provides INT, func accepts {paramType.Name}";
                    }
                    break;
                case VariableType.Double:
                    if (!paramType.IsAssignableFrom(typeof(double)))
                    {
                        errorMsg = $"TaskDef provides DOUBLE, func accepts {paramType.Name}";
                    }
                    break;
                case VariableType.Str:
                    if (!paramType.IsAssignableFrom(typeof(string)))
                    {
                        errorMsg = $"TaskDef provides STRING, func accepts {paramType.Name}";
                    }
                    break;
                case VariableType.Bool:
                    if (!paramType.IsAssignableFrom(typeof(bool)))
                    {
                        errorMsg = $"TaskDef provides BOOL, func accepts {paramType.Name}";
                    }
                    break;
                case VariableType.Bytes:
                    if (!paramType.IsAssignableFrom(typeof(byte[])))
                    {
                        errorMsg = $"TaskDef provides BYTES, func accepts {paramType.Name}";
                    }
                    break;
                case VariableType.JsonArr:
                case VariableType.JsonObj:
                    _logger?.LogInformation($"Will use Jackson to deserialize Json into {paramType.Name}");
                    break;
                case VariableType.Null:
                default:
                    throw new Exception("Not possible");
            }

            if (!string.IsNullOrEmpty(errorMsg))
            {
                errorMsg = $"Invalid assignment for var {paramName}: {errorMsg}";
                throw new LHTaskSchemaMismatchException(errorMsg);
            }
        }
    }
}
